src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs

```diff
@@ -24,33 +24,35 @@ public HttpClientTransport(HttpClient client = null)
 
         public readonly static HttpClientTransport Shared = new HttpClientTransport();
 
-        public sealed override HttpMessage CreateMessage(IServiceProvider services, CancellationToken cancellation)
-            => new Message(cancellation);
+        public sealed override HttpPipelineRequest CreateRequest(IServiceProvider services)
+            => new PipelineRequest();
 
-        public sealed override async Task ProcessAsync(HttpMessage message)
+        public sealed override async Task ProcessAsync(HttpPipelineContext pipelineContext)
         {
-            var httpTransportMessage = message as Message;
-            if (httpTransportMessage == null) throw new InvalidOperationException("the message is not compatible with the transport");
+            var pipelineRequest = pipelineContext.Request as PipelineRequest;
+            if (pipelineRequest == null) throw new InvalidOperationException("the message is not compatible with the transport");
 
-            HttpRequestMessage httpRequest = httpTransportMessage.BuildRequestMessage();
-
-            HttpResponseMessage responseMessage = await ProcessCoreAsync(message.Cancellation, httpRequest).ConfigureAwait(false);
-            httpTransportMessage.ProcessResponseMessage(responseMessage);
+            using (HttpRequestMessage httpRequest = pipelineRequest.BuildRequestMessage(pipelineContext.Cancellation))
+            {
+                HttpResponseMessage responseMessage = await ProcessCoreAsync(pipelineContext.Cancellation, httpRequest).ConfigureAwait(false);
+                pipelineContext.Response = new PipelineResponse(responseMessage);
```

Comment:
Respectfully, I would suggest a different design for this.   The side effect of modifying a parameter is something that I'd advocate trying to avoid.  In my opinion, it hurts readability and makes it more difficult to reason about what the method is doing.  In this case, we have no document comments decorating the method to hint to callers that we're changing the state of something they've given us.

Personally, I'd recommend having the method take a `PipelineRequest` and return a `PipelineResponse` so that its contract is clear to callers by its signature, with no implicit changes to parameters.

Reply:
I see your perspective. Couple comments here:

1. Returning Task<T> allocates on every call even when it never goes async - we can fix this by returning ValueTask<T> 
2. Policies need more than just request, cancellation token and options are also on context but we can keep passing that in

@KrzysztofCwalina thoughts?

Reply:
I think the policies should have access to the whole message/context. The "user" code (the SDK method using the pipeline) should use request/response. But this is based on intuition, not that I thought it thorough. In other words, not strong opinion on this yet.

Having said that, if all our APIs deal with request/response, I am not sure what the context would be for.

Reply:
I don't want to be the person blocking this PR, but I remain concerned with the approach of having a method that takes a parameter and makes modifications to that parameter as a blessed pattern for our core architecture.   The side-effect nature of that sequence makes it incredibly difficult to reason about the method without drilling into the implementation and goes against normal expectations.

I know that we're being very aware and restrained of our use in allocations, but I'd really be more comfortable here if we did something to make this clear via return value or using a `ref` or `out` signature for the response in the method itself.   At the very least, can we please annotate this with some XML doc comments and call out the side effect in some `<remarks>`?

@pakrym, @KrzysztofCwalina -  If you both are sure that this is the pattern that you want, I'll update to remove the "please change" blocker, but this just feels wrong to me.

Reply:
@jsquire I'm a bit less concerned about the pattern and I think it allows for the most flexibility:

There is some benefit for pipeline invocations to be able to see the "current" response. E.g. Auth policy might check if it needs to re-authenticate after a retry based on the last response etc.

>  The side-effect nature of that sequence makes it incredibly difficult to reason about the method without drilling into the implementation and goes against normal expectations.

I'm not sure I understand this part. For me, the reasoning is the following, 

1. Policies set up the request and call the next policy
1. Transport policy always sets the response
1. Policies get to take a look at response while the stack is unwinding and optionally repeat from 1.

Reply:
As a caller, when I send a parameter to a method, my expectation is that the method does not modify my parameter.   Unless the method denotes otherwise, typically by use of `ref` or `out` annotation, the expectation is that the caller owns the parameter; the method is a consumer and should have access to read an inspect it, but should not make changes to it.   Doing so violates that principal of "don't change things that you don't own" and runs against the expectation of a caller, potentially being a source of confusion.  

If the transport needs to modify the context, it should create a copy that it owns, set the response, and pass that copy to the pipeline so that the policies that follow see the transport copy and the caller's original remains unchanged.  The response, if interesting to the caller, can be returned from the method.

That said, I acknowledge that we're potentially in a performance-critical path and may have need to bend best practices to avoid allocations or hit other performance goals.  However, I feel strongly that it is worth at least giving the pattern full consideration and, if we decide that the trade-off is needed, make sure that we document to callers that we're doing something that is outside of the normal expectations.

Reply:
> As a caller, when I send a parameter to a method, my expectation is that the method does not modify my parameter.

I think you are overgeneralizing here, there are tons of mutable types and methods that mutate objects passed to them in .NET. In addition the pattern itself is not new, take a look at https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-2.2 that does almost the same things from the server side and seems to be working well.

Having things be immutable is nice in theory but from my experience, things fall apart when it comes to performance with current JIT and GC the hit is very substantial. In addition, we are not going to use the main benefit of immutable types - thread safety.

I would also like to note that this PR doesn't set things in stone if we see that code is confusing or doesn't work like expected we are free to iterate on it more.

Reply:
Respectfully, I feel the generalization is warranted.  Again, to be clear, I understand the pattern that we're using here and acknowledge that there's a place for it.  My goal is to ensure that it's the one chosen after we've considered the trade-offs and to be sure that we're properly documenting the expectations.    

If we feel that the side effects are warranted to keep to performance and resource goals, I've no objection.   My ask is simply that we make that clear by doc comment annotation of the method.